config.org
#+TITLE: Josh's Emacs Config
#+AUTHOR: Josh Porter
#+STARTUP: showeverything

* Table of Contents :toc:
- [[#package-manager][Package Manager]]
- [[#general-settings][General Settings]]
  - [[#better-scrolling][Better Scrolling]]
  - [[#better-mouse-scrolling][Better Mouse Scrolling]]
- [[#ui][UI]]
  - [[#fonts][Fonts]]
  - [[#theme][Theme]]
  - [[#modeline][Modeline]]
  - [[#transparency][Transparency]]
  - [[#line-numbers][Line Numbers]]
  - [[#neotree][Neotree]]
  - [[#dashboard][Dashboard]]
- [[#lsp-and-completion][LSP and Completion]]
  - [[#eglot][Eglot]]
  - [[#corfu][Corfu]]
  - [[#languages][Languages]]
  - [[#treesitter][Treesitter]]
- [[#evil][Evil]]
- [[#additional-packages][Additional Packages]]
  - [[#ivy][Ivy]]
  - [[#counsel][Counsel]]
  - [[#whichkey][Whichkey]]
  - [[#magit][Magit]]
  - [[#git-gutter][Git Gutter]]
  - [[#gcmh][GCMH]]
  - [[#projectile][Projectile]]
  - [[#vterm][Vterm]]
- [[#keybinds][Keybinds]]
  - [[#general][General]]
- [[#org-mode][Org Mode]]
  - [[#org-mode-bullets][Org Mode Bullets]]
  - [[#table-of-contents][Table of Contents]]

* Package Manager
#+begin_src emacs-lisp

(require 'package)
(add-to-list 'package-archives '("gnu"   . "https://elpa.gnu.org/packages/"))
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(package-initialize)

(unless package-archive-contents
  (package-refresh-contents))

(unless (package-installed-p 'use-package)
  (package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)

#+end_src

* General Settings
#+begin_src emacs-lisp

(setq inhibit-startup-message t)            ; Disable startup message
(setq auto-save-default nil)                ; Disable autosaves
(setq backup-inhibited t)                   ; Disable backups
(tool-bar-mode -1)                          ; Disable tool bar
(menu-bar-mode -1)                          ; Disable menu bar
(scroll-bar-mode -1)                        ; Disable scroll bar
(setq ring-bell-function 'ignore)           ; Disable bell
(setq-default tab-width 2)
(setq-default evil-shift-width tab-width)
(electric-pair-mode)

#+end_src

** Better Scrolling
#+begin_src emacs-lisp

(setq scroll-step 1)
(setq scroll-conservatively 101)
(setq fast-but-imprecise-scrolling nil)
(setq jit-lock-defer-time 0)

#+end_src

** Better Mouse Scrolling
#+begin_src emacs-lisp

(use-package mwheel
  :ensure nil
  :config
  (setq mouse-wheel-scroll-amount '(3 ((shift) . 1) ((control) . nil)))
  (setq mouse-wheel-progressive-speed nil))
  
#+end_src

* UI
** Fonts
#+begin_src emacs-lisp

(add-to-list 'default-frame-alist
	     '(font . "JetBrainsMono NF-10"))

#+end_src

** Theme
#+begin_src emacs-lisp

(use-package doom-themes
  :config
  (load-theme 'doom-tokyo-night t))

#+end_src

** Modeline
#+begin_src emacs-lisp

(use-package doom-modeline
  :init (doom-modeline-mode 1))

;; Show time and battery information on modeline
(display-time-mode 1)
(display-battery-mode 1)

#+end_src

** Transparency
#+begin_src emacs-lisp

(set-frame-parameter (selected-frame) 'alpha '(98 . 98))
(add-to-list 'default-frame-alist '(alpha . (98 . 98)))

#+end_src

** Line Numbers
#+begin_src emacs-lisp

(use-package display-line-numbers
  :ensure nil
  :hook (prog-mode . display-line-numbers-mode)
  :config
  (setq-default display-line-numbers-width 3))

#+end_src

** Neotree
#+begin_src emacs-lisp

(use-package all-the-icons)

#+end_src

#+begin_src emacs-lisp

(use-package neotree
	:config
	(setq neo-theme 'nerd))

#+end_src


** Dashboard
#+begin_src emacs-lisp

(use-package dashboard
  :config
  (dashboard-setup-startup-hook)
  (setq dashboard-startup-banner 'logo)
  (setq dashboard-banner-logo-title "Josh's Emacs")
	(setq dashboard-items '((recents  . 5)
                        (bookmarks . 5)
                        (projects . 5)
                        (agenda . 5)
                        (registers . 5))))
#+end_src

* LSP and Completion
** Eglot
#+begin_src emacs-lisp

(use-package eglot)

#+end_src

** Corfu
#+begin_src emacs-lisp

(use-package corfu
  ;; Optional customizations
  :custom
  ;; (corfu-cycle t)                ;; Enable cycling for `corfu-next/previous'
  (corfu-auto t)                    ;; Enable auto completion
  ;; (corfu-separator ?\s)          ;; Orderless field separator
  ;; (corfu-quit-at-boundary nil)   ;; Never quit at completion boundary
  ;; (corfu-quit-no-match nil)      ;; Never quit, even if there is no match
  ;; (corfu-preview-current nil)    ;; Disable current candidate preview
  ;; (corfu-preselect 'prompt)      ;; Preselect the prompt
  ;; (corfu-on-exact-match nil)     ;; Configure handling of exact matches
  ;; (corfu-scroll-margin 5)        ;; Use scroll margin

  ;; Enable Corfu only for certain modes.
  ;; :hook ((prog-mode . corfu-mode)
  ;;        (shell-mode . corfu-mode)
  ;;        (eshell-mode . corfu-mode))

  ;; Recommended: Enable Corfu globally.
  ;; This is recommended since Dabbrev can be used globally (M-/).
  ;; See also `corfu-exclude-modes'.
  :init
  (global-corfu-mode)
  (corfu-popupinfo-mode))

#+end_src

** Languages
*** Go
#+begin_src emacs-lisp

(use-package go-mode)

#+end_src

*** Rust
#+begin_src emacs-lisp

(use-package rust-mode
  :bind (("<f6>" . josh/rust-format-buffer))
	:config
	(require 'rust-rustfmt)
	(defun josh/rust-format-buffer ()
	  (interactive)
		(rust-format-buffer)
		(save-buffer)))

#+end_src

** Treesitter
#+begin_src emacs-lisp

(use-package tree-sitter
  :after tree-sitter-langs
  :config
  (global-tree-sitter-mode)
  (add-hook 'tree-sitter-after-on-hook #'tree-sitter-hl-mode))

(use-package tree-sitter-langs)

#+end_src

* Evil
#+begin_src emacs-lisp

(use-package evil
  :init
  (setq evil-want-integration t)
  (setq evil-want-keybinding nil)
  (setq evil-want-C-u-scroll t)
  :config
  (evil-mode 1))

(use-package evil-collection
  :after evil
  :config
  (evil-collection-init))

#+end_src

* Additional Packages
** Ivy
#+begin_src emacs-lisp

(use-package ivy
  :diminish
  :bind (("C-s" . swiper)
         :map ivy-minibuffer-map
         ("TAB" . ivy-alt-done)
         ("C-l" . ivy-alt-done)
         ("C-j" . ivy-next-line)
         ("C-k" . ivy-previous-line)
         :map ivy-switch-buffer-map
         ("C-k" . ivy-previous-line)
         ("C-l" . ivy-done)
         ("C-d" . ivy-switch-buffer-kill)
         :map ivy-reverse-i-search-map
         ("C-k" . ivy-previous-line)
         ("C-d" . ivy-reverse-i-search-kill))
  :config
  (ivy-mode 1))

#+end_src

** Counsel
#+begin_src emacs-lisp

(use-package counsel
             :bind(("M-x" . counsel-M-x)
                   ("C-x b" . counsel-ibuffer)
                   ("C-x C-f" . counsel-find-file)
		   ("C-M-j" . counsel-switch-buffer)
                   :map minibuffer-local-map
                   ("C-r" . counsel-minibuffer-history))
             :config
             (setq ivy-initial-inputs-alist nil))

#+end_src

** Whichkey
#+begin_src emacs-lisp

(use-package which-key
  :init
    (which-key-mode 1)
  :config
  (setq which-key-side-window-location 'bottom
	  which-key-sort-order #'which-key-key-order-alpha
	  which-key-sort-uppercase-first nil
	  which-key-add-column-padding 1
	  which-key-max-display-columns nil
	  which-key-min-display-lines 6
	  which-key-side-window-slot -10
	  which-key-side-window-max-height 0.25
	  which-key-idle-delay 0.8
	  which-key-max-description-length 25
	  which-key-allow-imprecise-window-fit t
	  which-key-separator " → " ))

#+end_src

** Magit
#+begin_src emacs-lisp

(use-package magit)

#+end_src

** Git Gutter
#+begin_src emacs-lisp

(use-package git-gutter
  :hook (prog-mode . git-gutter-mode)
  :config
  (setq git-gutter:update-interval 0.02))

(use-package git-gutter-fringe
  :config
  (define-fringe-bitmap 'git-gutter-fr:added [224] nil nil '(center repeated))
  (define-fringe-bitmap 'git-gutter-fr:modified [224] nil nil '(center repeated))
  (define-fringe-bitmap 'git-gutter-fr:deleted [224] nil nil '(center repeated)))

#+end_src

** GCMH
#+begin_src emacs-lisp

(use-package gcmh
  :demand t
  :config
  (setq gcmh-low-cons-threshold (* 16 1024 1024))
  (gcmh-mode +1))

#+end_src

** Projectile
#+begin_src emacs-lisp

(use-package projectile
  :diminish projectile-mode
  :config (projectile-mode)
  :custom ((projectile-completion-system 'ivy))
  :bind-keymap
  ("C-c p" . projectile-command-map)
  :init
  (when (file-directory-p "~/dev")
    (setq projecttile-project-search-path '("~/dev")))
  (setq projectile-switch-project-action #'projectile-dired))

#+end_src

** Vterm
#+begin_src emacs-lisp

(use-package vterm
  :config
  (setq shell-file-name "/usr/bin/zsh"))

(use-package vterm-toggle
  :after vterm
  :config
  (setq vterm-toggle-fullscreen-p 1)
  (setq vterm-toggle-scope 'project)
  (add-to-list 'display-buffer-alist
               '((lambda (buffer-or-name _)
                     (let ((buffer (get-buffer buffer-or-name)))
                       (with-current-buffer buffer
                         (or (equal major-mode 'vterm-mode)
                             (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
                  (display-buffer-reuse-window display-buffer-at-bottom)
                  ;;(display-buffer-reuse-window display-buffer-in-direction)
                  ;;display-buffer-in-direction/direction/dedicated is added in emacs27
                  ;;(direction . bottom)
                  ;;(dedicated . t) ;dedicated is supported in emacs27
                  (reusable-frames . visible)
                  (window-height . 0.3))))

#+end_src

* Keybinds
** General
#+begin_src emacs-lisp

(use-package general
  :config
  (general-evil-setup)

  (general-create-definer josh/leader-keys
    :states '(normal insert visual emacs)
    :keymaps 'override
    :prefix "SPC"
    :global-prefix "M-SPC")

  (josh/leader-keys
    "cc" '(comment-or-uncomment-region :wk "comment line"))

  (josh/leader-keys
    "." '(counsel-find-file :wk "Find file")
    "f c" '((lambda () (interactive) (find-file "~/.emacs.d/config.org")) :wk "Edit emacs config")
    "f r" '(counsel-recentf :wk "Find recent files"))

  (josh/leader-keys
    "t" '(:ignore t :wk "Toggle")
    "t n" '(neotree-toggle :wk "neotree-toggle")
    "t t" '(counsel-load-theme :wk "switch theme")
    "t v" '(vterm-toggle :wk "Toggle vterm"))

  (josh/leader-keys
    "b" '(:ignore t :wk "buffer")
    "b b" '(counsel-switch-buffer :wk "Switch buffer")
    "b i" '(ibuffer :wk "Ibuffer")
    "b k" '(kill-this-buffer :wk "Kill this buffer")
    "b n" '(next-buffer :wk "Next buffer")
    "b p" '(previous-buffer :wk "Previous buffer")
    "b r" '(revert-buffer :wk "Reload buffer"))

  (josh/leader-keys
    "w" '(:ignore t :wk "Windows")
    ;; Window splits
    "w c" '(evil-window-delete :wk "Close window")
    "w n" '(evil-window-new :wk "New window")
    "w s" '(evil-window-split :wk "Horizontal split window")
    "w v" '(evil-window-vsplit :wk "Vertical split window")
    ;; Window motions
    "w h" '(evil-window-left :wk "Window left")
    "w j" '(evil-window-down :wk "Window down")
    "w k" '(evil-window-up :wk "Window up")
    "w l" '(evil-window-right :wk "Window right")
    "w w" '(evil-window-next :wk "Goto next window"))
)

#+end_src

* Org Mode
** Org Mode Bullets
#+begin_src emacs-lisp

(use-package org-bullets
  :after org
  :hook (org-mode . org-bullets-mode))

#+end_src

** Table of Contents
#+begin_src emacs-lisp

(use-package toc-org
    :commands toc-org-enable
    :init (add-hook 'org-mode-hook 'toc-org-enable))

#+end_src


config.el
(require 'package)
(add-to-list 'package-archives '("gnu"   . "https://elpa.gnu.org/packages/"))
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(package-initialize)

(unless package-archive-contents
  (package-refresh-contents))

(unless (package-installed-p 'use-package)
  (package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)

(setq inhibit-startup-message t)            ; Disable startup message
(setq auto-save-default nil)                ; Disable autosaves
(setq backup-inhibited t)                   ; Disable backups
(tool-bar-mode -1)                          ; Disable tool bar
(menu-bar-mode -1)                          ; Disable menu bar
(scroll-bar-mode -1)                        ; Disable scroll bar
(setq ring-bell-function 'ignore)           ; Disable bell
(setq-default tab-width 2)
(setq-default evil-shift-width tab-width)
(electric-pair-mode)

(setq scroll-step 1)
(setq scroll-conservatively 101)
(setq fast-but-imprecise-scrolling nil)
(setq jit-lock-defer-time 0)

(use-package mwheel
  :ensure nil
  :config
  (setq mouse-wheel-scroll-amount '(3 ((shift) . 1) ((control) . nil)))
  (setq mouse-wheel-progressive-speed nil))

(add-to-list 'default-frame-alist
	     '(font . "JetBrainsMono NF-10"))

(use-package doom-themes
  :config
  (load-theme 'doom-tokyo-night t))

(use-package doom-modeline
  :init (doom-modeline-mode 1))

;; Show time and battery information on modeline
(display-time-mode 1)
(display-battery-mode 1)

(set-frame-parameter (selected-frame) 'alpha '(98 . 98))
(add-to-list 'default-frame-alist '(alpha . (98 . 98)))

(use-package display-line-numbers
  :ensure nil
  :hook (prog-mode . display-line-numbers-mode)
  :config
  (setq-default display-line-numbers-width 3))

(use-package all-the-icons)

(use-package neotree
	:config
	(setq neo-theme 'nerd))

(use-package dashboard
  :config
  (dashboard-setup-startup-hook)
  (setq dashboard-startup-banner 'logo)
  (setq dashboard-banner-logo-title "Josh's Emacs")
	(setq dashboard-items '((recents  . 5)
                        (bookmarks . 5)
                        (projects . 5)
                        (agenda . 5)
                        (registers . 5))))

(use-package eglot)

(use-package corfu
  ;; Optional customizations
  :custom
  ;; (corfu-cycle t)                ;; Enable cycling for `corfu-next/previous'
  (corfu-auto t)                    ;; Enable auto completion
  ;; (corfu-separator ?\s)          ;; Orderless field separator
  ;; (corfu-quit-at-boundary nil)   ;; Never quit at completion boundary
  ;; (corfu-quit-no-match nil)      ;; Never quit, even if there is no match
  ;; (corfu-preview-current nil)    ;; Disable current candidate preview
  ;; (corfu-preselect 'prompt)      ;; Preselect the prompt
  ;; (corfu-on-exact-match nil)     ;; Configure handling of exact matches
  ;; (corfu-scroll-margin 5)        ;; Use scroll margin

  ;; Enable Corfu only for certain modes.
  ;; :hook ((prog-mode . corfu-mode)
  ;;        (shell-mode . corfu-mode)
  ;;        (eshell-mode . corfu-mode))

  ;; Recommended: Enable Corfu globally.
  ;; This is recommended since Dabbrev can be used globally (M-/).
  ;; See also `corfu-exclude-modes'.
  :init
  (global-corfu-mode)
  (corfu-popupinfo-mode))

(use-package go-mode)

(use-package rust-mode
  :bind (("<f6>" . josh/rust-format-buffer))
	:config
	(require 'rust-rustfmt)
	(defun josh/rust-format-buffer ()
	  (interactive)
		(rust-format-buffer)
		(save-buffer)))

(use-package tree-sitter
  :after tree-sitter-langs
  :config
  (global-tree-sitter-mode)
  (add-hook 'tree-sitter-after-on-hook #'tree-sitter-hl-mode))

(use-package tree-sitter-langs)

(use-package evil
  :init
  (setq evil-want-integration t)
  (setq evil-want-keybinding nil)
  (setq evil-want-C-u-scroll t)
  :config
  (evil-mode 1))

(use-package evil-collection
  :after evil
  :config
  (evil-collection-init))

(use-package ivy
  :diminish
  :bind (("C-s" . swiper)
         :map ivy-minibuffer-map
         ("TAB" . ivy-alt-done)
         ("C-l" . ivy-alt-done)
         ("C-j" . ivy-next-line)
         ("C-k" . ivy-previous-line)
         :map ivy-switch-buffer-map
         ("C-k" . ivy-previous-line)
         ("C-l" . ivy-done)
         ("C-d" . ivy-switch-buffer-kill)
         :map ivy-reverse-i-search-map
         ("C-k" . ivy-previous-line)
         ("C-d" . ivy-reverse-i-search-kill))
  :config
  (ivy-mode 1))

(use-package counsel
             :bind(("M-x" . counsel-M-x)
                   ("C-x b" . counsel-ibuffer)
                   ("C-x C-f" . counsel-find-file)
		   ("C-M-j" . counsel-switch-buffer)
                   :map minibuffer-local-map
                   ("C-r" . counsel-minibuffer-history))
             :config
             (setq ivy-initial-inputs-alist nil))

(use-package which-key
  :init
    (which-key-mode 1)
  :config
  (setq which-key-side-window-location 'bottom
	  which-key-sort-order #'which-key-key-order-alpha
	  which-key-sort-uppercase-first nil
	  which-key-add-column-padding 1
	  which-key-max-display-columns nil
	  which-key-min-display-lines 6
	  which-key-side-window-slot -10
	  which-key-side-window-max-height 0.25
	  which-key-idle-delay 0.8
	  which-key-max-description-length 25
	  which-key-allow-imprecise-window-fit t
	  which-key-separator " → " ))

(use-package magit)

(use-package git-gutter
  :hook (prog-mode . git-gutter-mode)
  :config
  (setq git-gutter:update-interval 0.02))

(use-package git-gutter-fringe
  :config
  (define-fringe-bitmap 'git-gutter-fr:added [224] nil nil '(center repeated))
  (define-fringe-bitmap 'git-gutter-fr:modified [224] nil nil '(center repeated))
  (define-fringe-bitmap 'git-gutter-fr:deleted [224] nil nil '(center repeated)))

(use-package gcmh
  :demand t
  :config
  (setq gcmh-low-cons-threshold (* 16 1024 1024))
  (gcmh-mode +1))

(use-package projectile
  :diminish projectile-mode
  :config (projectile-mode)
  :custom ((projectile-completion-system 'ivy))
  :bind-keymap
  ("C-c p" . projectile-command-map)
  :init
  (when (file-directory-p "~/dev")
    (setq projecttile-project-search-path '("~/dev")))
  (setq projectile-switch-project-action #'projectile-dired))

(use-package vterm
  :config
  (setq shell-file-name "/usr/bin/zsh"))

(use-package vterm-toggle
  :after vterm
  :config
  (setq vterm-toggle-fullscreen-p 1)
  (setq vterm-toggle-scope 'project)
  (add-to-list 'display-buffer-alist
               '((lambda (buffer-or-name _)
                     (let ((buffer (get-buffer buffer-or-name)))
                       (with-current-buffer buffer
                         (or (equal major-mode 'vterm-mode)
                             (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
                  (display-buffer-reuse-window display-buffer-at-bottom)
                  ;;(display-buffer-reuse-window display-buffer-in-direction)
                  ;;display-buffer-in-direction/direction/dedicated is added in emacs27
                  ;;(direction . bottom)
                  ;;(dedicated . t) ;dedicated is supported in emacs27
                  (reusable-frames . visible)
                  (window-height . 0.3))))

(use-package general
  :config
  (general-evil-setup)

  (general-create-definer josh/leader-keys
    :states '(normal insert visual emacs)
    :keymaps 'override
    :prefix "SPC"
    :global-prefix "M-SPC")

  (josh/leader-keys
    "cc" '(comment-or-uncomment-region :wk "comment line"))

  (josh/leader-keys
    "." '(counsel-find-file :wk "Find file")
    "f c" '((lambda () (interactive) (find-file "~/.emacs.d/config.org")) :wk "Edit emacs config")
    "f r" '(counsel-recentf :wk "Find recent files"))

  (josh/leader-keys
    "t" '(:ignore t :wk "Toggle")
    "t n" '(neotree-toggle :wk "neotree-toggle")
    "t t" '(counsel-load-theme :wk "switch theme")
    "t v" '(vterm-toggle :wk "Toggle vterm"))

  (josh/leader-keys
    "b" '(:ignore t :wk "buffer")
    "b b" '(counsel-switch-buffer :wk "Switch buffer")
    "b i" '(ibuffer :wk "Ibuffer")
    "b k" '(kill-this-buffer :wk "Kill this buffer")
    "b n" '(next-buffer :wk "Next buffer")
    "b p" '(previous-buffer :wk "Previous buffer")
    "b r" '(revert-buffer :wk "Reload buffer"))

  (josh/leader-keys
    "w" '(:ignore t :wk "Windows")
    ;; Window splits
    "w c" '(evil-window-delete :wk "Close window")
    "w n" '(evil-window-new :wk "New window")
    "w s" '(evil-window-split :wk "Horizontal split window")
    "w v" '(evil-window-vsplit :wk "Vertical split window")
    ;; Window motions
    "w h" '(evil-window-left :wk "Window left")
    "w j" '(evil-window-down :wk "Window down")
    "w k" '(evil-window-up :wk "Window up")
    "w l" '(evil-window-right :wk "Window right")
    "w w" '(evil-window-next :wk "Goto next window"))
)

(use-package org-bullets
  :after org
  :hook (org-mode . org-bullets-mode))

(use-package toc-org
    :commands toc-org-enable
    :init (add-hook 'org-mode-hook 'toc-org-enable))
